OneSignalSDK/onesignal/src/main/java/com/onesignal/OSSessionManager.java

```diff
@@ -0,0 +1,295 @@
+    }
+
+    /**
+     * Set the notifications ids that influenced the session
+     */
+    @NonNull protected JSONArray getLastNotificationsReceivedIds() {
+        JSONArray notificationsReceived = OutcomesUtils.getLastNotificationsReceivedData();
+        JSONArray notificationsIds = new JSONArray();
+
+        long attributionWindow = OutcomesUtils.getIndirectAttributionWindow() * 60 * 1_000L;
+        long currentTime = System.currentTimeMillis();
+        for (int i = 0; i < notificationsReceived.length(); i++) {
+            try {
+                JSONObject jsonObject = notificationsReceived.getJSONObject(i);
+                long time = jsonObject.getLong(OutcomesUtils.TIME);
+                long difference = currentTime - time;
+
+                if (difference <= attributionWindow) {
+                    String notificationId = jsonObject.getString(OutcomesUtils.NOTIFICATION_ID);
+                    notificationsIds.put(notificationId);
+                }
+            } catch (JSONException e) {
+                OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "From getting notification from array:JSON Failed.", e);
+            }
+        }
+
+        return notificationsIds;
+    }
+
+    // Get the current session based on state + if outcomes features are enabled.
+    @NonNull SessionResult getSessionResult() {
+        if (session.isDirect()) {
+            if (OutcomesUtils.isDirectSessionEnabled()) {
+                JSONArray directNotificationIds = new JSONArray().put(directNotificationId);
+                return SessionResult.Builder.newInstance()
+                        .setNotificationIds(directNotificationIds)
+                        .setSession(Session.DIRECT)
+                        .build();
+            }
+        } else if (session.isIndirect()) {
+            if (OutcomesUtils.isIndirectSessionEnabled()) {
+                return SessionResult.Builder.newInstance()
+                        .setNotificationIds(indirectNotificationIds)
+                        .setSession(Session.INDIRECT)
+                        .build();
+            }
+        } else if (OutcomesUtils.isUnattributedSessionEnabled()) {
+            return SessionResult.Builder.newInstance()
+                    .setSession(Session.UNATTRIBUTED)
+                    .build();
+        }
+
+        return SessionResult.Builder.newInstance()
+                .setSession(Session.DISABLED)
+                .build();
+    }
+
+    /**
+     * Attempt to override the current session before the 30 second session minimum
+     * This should only be done in a upward direction:
+     *   * UNATTRIBUTED can become INDIRECT or DIRECT
+     *   * INDIRECT can become DIRECT
+     *   * DIRECT can become DIRECT
+     */
+    void attemptSessionUpgrade() {
+        // We will try to override any session with DIRECT
+        if (OneSignal.getAppEntryState().isNotificationClick()) {
+            setSession(Session.DIRECT, directNotificationId, null);
```

Comment:
cant we depend on the directNotificationId to be set? and to pass as param OneSignal.getAppEntryState().isNotificationClick() boolean? to remove this static dependency and keep only composition relation from OneSignal to SessionManager

Reply:
If `getAppEntryState().isNotificationClick()` is `true` we would have already had to set the `directNotificationId`

Reply:
Let's leave this as is for now. We can refactor how the dependency on `appEntryState` is structured later